File: 02_Data_Structures/stacks_queues/stack/implementations/operations.cpp
// File: operations.cpp
// Topic: Stack (Array/Linked-List Implementation)
// Description:
//   Implementation of a stack supporting push, pop, top, size, and isEmpty.
//   Includes both array-based and linked-list based stack structures.
//   Demonstrates LIFO (Last In First Out) behavior.
//
// Algorithm Choice:
//   Array-based stack gives O(1) operations with fixed capacity.
//   Linked-list stack allows dynamic growth without overflow.
//
// Complexity:
//   Time: O(1) for push/pop/top
//   Space: O(n) for n elements in the stack
//
// Variations:
//   - Dynamic array stack (auto-resizing)
//   - Stack using two queues
//   - Stack using deque

#include <iostream>
using namespace std;

// -----------------------------
// ARRAY-BASED STACK
// -----------------------------
class ArrayStack {
private:
    int *arr;
    int topIndex;
    int capacity;

public:
    ArrayStack(int cap = 1000) {
        capacity = cap;
        arr = new int[capacity];
        topIndex = -1;
    }

    void push(int x) {
        if (topIndex == capacity - 1) {
            cout << "Stack Overflow\n";
            return;
        }
        arr[++topIndex] = x;
    }

    void pop() {
        if (topIndex == -1) {
            cout << "Stack Underflow\n";
            return;
        }
        topIndex--;
    }

    int top() {
        if (topIndex == -1) {
            cout << "Stack Empty\n";
            return -1;
        }
        return arr[topIndex];
    }

    bool isEmpty() {
        return (topIndex == -1);
    }

    int size() {
        return topIndex + 1;
    }
};

// -----------------------------
// LINKED LIST BASED STACK
// -----------------------------
struct Node {
    int data;
    Node* next;
    Node(int val) : data(val), next(NULL) {}
};

class LinkedStack {
private:
    Node* head;
    int count;

public:
    LinkedStack() {
        head = NULL;
        count = 0;
    }

    void push(int x) {
        Node* newNode = new Node(x);
        newNode->next = head;
        head = newNode;
        count++;
    }

    void pop() {
        if (head == NULL) {
            cout << "Stack Underflow\n";
            return;
        }
        Node* temp = head;
        head = head->next;
        delete temp;
        count--;
    }

    int top() {
        if (head == NULL) {
            cout << "Stack Empty\n";
            return -1;
        }
        return head->data;
    }

    bool isEmpty() {
        return head == NULL;
    }

    int size() {
        return count;
    }
};

// -----------------------------
// DRIVER CODE (FOR DEMO)
// -----------------------------
int main() {
    cout << "Array Stack demo:\n";
    ArrayStack st(5);
    st.push(10);
    st.push(20);
    st.push(30);
    cout << "Top: " << st.top() << "\n"; // 30
    st.pop();
    cout << "Top after pop: " << st.top() << "\n"; // 20

    cout << "\nLinked Stack demo:\n";
    LinkedStack lst;
    lst.push(5);
    lst.push(15);
    lst.push(25);
    cout << "Top: " << lst.top() << "\n"; // 25
    lst.pop();
    cout << "Top after pop: " << lst.top() << "\n"; // 15

    return 0;
}

File: 02_Data_Structures/stacks_queues/stack/implementations/operations.py
# File: operations.py
# Topic: Stack (Array-based and Linked-list based implementation)
# Description:
#   A complete stack implementation demonstrating LIFO behavior.
#   Supports push, pop, top/peek, size, and is_empty operations.
#
# Algorithm Choice:
#   - Array-based stack: fast O(1) operations using Python list.
#   - Linked-list based stack: dynamic memory growth without overflow.
#
# Time Complexity:
#   Push: O(1)
#   Pop: O(1)
#   Peek: O(1)
#   is_empty: O(1)
#   size: O(1)
#
# Variations:
#   - Dynamic array expansion (handled internally by Python list)
#   - Stack implemented using two queues
#   - Stack implemented using deque (collections.deque)

# --------------------------
# ARRAY-BASED STACK
# --------------------------
class ArrayStack:
    def __init__(self):
        self.arr = []

    def push(self, x):
        self.arr.append(x)  # O(1)

    def pop(self):
        if not self.arr:
            print("Stack Underflow")
            return None
        return self.arr.pop()  # O(1)

    def top(self):
        if not self.arr:
            print("Stack Empty")
            return None
        return self.arr[-1]  # O(1)

    def is_empty(self):
        return len(self.arr) == 0

    def size(self):
        return len(self.arr)


# --------------------------
# LINKED LIST BASED STACK
# --------------------------
class Node:
    def __init__(self, val):
        self.data = val
        self.next = None


class LinkedStack:
    def __init__(self):
        self.head = None
        self.count = 0

    def push(self, x):
        new_node = Node(x)
        new_node.next = self.head
        self.head = new_node
        self.count += 1

    def pop(self):
        if not self.head:
            print("Stack Underflow")
            return None

        value = self.head.data
        self.head = self.head.next
        self.count -= 1
        return value

    def top(self):
        if not self.head:
            print("Stack Empty")
            return None
        return self.head.data

    def is_empty(self):
        return self.head is None

    def size(self):
        return self.count


# --------------------------
# DEMO (Optional)
# --------------------------
if __name__ == "__main__":
    print("ArrayStack demo:")
    s = ArrayStack()
    s.push(10)
    s.push(20)
    print("Top:", s.top())  # 20
    s.pop()
    print("Top after pop:", s.top())  # 10

    print("\nLinkedStack demo:")
    ls = LinkedStack()
    ls.push(5)
    ls.push(15)
    print("Top:", ls.top())  # 15
    ls.pop()
    print("Top after pop:", ls.top())  # 5
